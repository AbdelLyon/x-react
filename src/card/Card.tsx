import type { JSX, ReactNode } from "react";
import { forwardRef } from "react";
import type {
  CardFooterProps,
  CardProps as NextUICardProps,
} from "@nextui-org/card";
import {
  Card as NextUICard,
  CardHeader,
  CardBody,
  CardFooter,
} from "@nextui-org/card";

interface CardProps extends NextUICardProps {
  header?: ReactNode;
  footer?: ReactNode;
  footerProps?: CardFooterProps;
  classNames?: Partial<Record<"base" | "header" | "body" | "footer", string>>;
}

export const Card = forwardRef<HTMLDivElement, CardProps>(
  (
    {
      children,
      shadow = "md",
      radius = "lg",
      fullWidth = false,
      isHoverable = false,
      isPressable = false,
      isBlurred = false,
      isFooterBlurred = false,
      isDisabled = false,
      disableAnimation = false,
      disableRipple = false,
      allowTextSelectionOnPress = false,
      classNames,
      header,
      footer,
      footerProps,
      ...props
    },
    ref,
  ): JSX.Element => {
    return (
      <NextUICard
        ref={ref}
        shadow={shadow}
        radius={radius}
        fullWidth={fullWidth}
        isHoverable={isHoverable}
        isPressable={isPressable}
        isBlurred={isBlurred}
        isFooterBlurred={isFooterBlurred}
        isDisabled={isDisabled}
        disableAnimation={disableAnimation}
        disableRipple={disableRipple}
        allowTextSelectionOnPress={allowTextSelectionOnPress}
        className={classNames?.base}
        {...props}
      >
        {header !== undefined && (
          <CardHeader className={classNames?.header}>{header}</CardHeader>
        )}
        <CardBody className={classNames?.body}>{children}</CardBody>
        {footer !== undefined && (
          <CardFooter className={classNames?.footer} {...footerProps}>
            {footer}
          </CardFooter>
        )}
      </NextUICard>
    );
  },
);

Card.displayName = "Card";
