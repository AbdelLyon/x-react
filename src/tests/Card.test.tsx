import { Card } from "@/card";
import { render, fireEvent } from "@testing-library/react";
import { describe, it, expect, vi } from "vitest";

describe("Card", (): void => {
  const defaultProps = {
    children: "Card Content",
    header: "Card Header",
    footer: "Card Footer",
  };

  it("renders with default props", (): void => {
    const { getByText } = render(<Card {...defaultProps} />);
    expect(getByText("Card Content")).toBeTruthy();
    expect(getByText("Card Header")).toBeTruthy();
    expect(getByText("Card Footer")).toBeTruthy();
  });

  it("applies custom classNames", (): void => {
    const customClassNames = {
      header: "custom-header",
      body: "custom-body",
      footer: "custom-footer",
    };

    const { container } = render(
      <Card {...defaultProps} classNames={customClassNames} />,
    );

    expect(container.querySelector(".custom-header")).toBeTruthy();
    expect(container.querySelector(".custom-body")).toBeTruthy();
    expect(container.querySelector(".custom-footer")).toBeTruthy();
  });

  it("handles press events", (): void => {
    const onPress = vi.fn();
    const { getByText } = render(
      <Card {...defaultProps} isPressable onPress={onPress} />,
    );

    const cardElement = getByText("Card Content").parentElement;
    if (cardElement) {
      fireEvent.click(cardElement);
      expect(onPress).toHaveBeenCalled();
    } else {
      throw new Error("Card element not found");
    }
  });

  it("renders without header and footer when not provided", (): void => {
    const { queryByText } = render(<Card>Card Content Only</Card>);
    expect(queryByText("Card Content Only")).toBeTruthy();
    expect(queryByText("Card Header")).toBeFalsy();
    expect(queryByText("Card Footer")).toBeFalsy();
  });

  it("applies visual props correctly", (): void => {
    const { container } = render(
      <Card
        shadow="lg"
        radius="sm"
        fullWidth
        isHoverable
        isBlurred
        isFooterBlurred
        isDisabled
      >
        Card Content
      </Card>,
    );

    const cardElement = container.firstChild;
    expect(cardElement).toHaveClass("shadow-large");
    expect(cardElement).toHaveClass("rounded-small");
    expect(cardElement).toHaveClass("w-full");
    expect(cardElement).toHaveClass("bg-background/80");
    expect(cardElement).toHaveClass("opacity-disabled");
  });

  it("forwards ref correctly", (): void => {
    const ref = { current: null };
    render(<Card ref={ref}>Card Content</Card>);
    expect(ref.current).toBeTruthy();
  });

  it("handles interaction props", (): void => {
    const { container } = render(
      <Card
        isPressable
        disableAnimation
        disableRipple
        allowTextSelectionOnPress
      >
        Interactive Card
      </Card>,
    );

    const cardElement = container.firstChild;
    expect(cardElement).toHaveClass("cursor-pointer");
  });
});
