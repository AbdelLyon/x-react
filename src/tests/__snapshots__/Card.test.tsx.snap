// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Composant Card > Snapshots > devrait correspondre au snapshot avec des classNames personnalisés 1`] = `
<div>
  <div
    class="flex flex-col relative overflow-hidden h-auto text-foreground box-border bg-content1 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium rounded-large transition-transform-background motion-reduce:transition-none"
    tabindex="-1"
  >
    <div
      class="flex p-3 z-10 w-full justify-start items-center shrink-0 overflow-inherit color-inherit subpixel-antialiased rounded-t-large custom-header"
    >
      En-tête de la Card
    </div>
    <div
      class="relative flex w-full p-3 flex-auto flex-col place-content-inherit align-items-inherit h-auto break-words text-left overflow-y-auto subpixel-antialiased custom-body"
    >
      Contenu de la Card
    </div>
    <div
      class="p-3 h-auto flex w-full items-center overflow-hidden color-inherit subpixel-antialiased rounded-b-large custom-footer"
    >
      Pied de la Card
    </div>
  </div>
</div>
`;

exports[`Composant Card > Snapshots > devrait correspondre au snapshot avec des props d'interaction 1`] = `
<div>
  <button
    class="flex flex-col relative overflow-hidden h-auto text-foreground box-border bg-content1 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium rounded-large cursor-pointer data-[pressed=true]:scale-[0.97] tap-highlight-transparent"
    role="button"
    type="button"
  >
    <div
      class="relative flex w-full p-3 flex-auto flex-col place-content-inherit align-items-inherit h-auto break-words text-left overflow-y-auto subpixel-antialiased"
    >
      Card Interactive
    </div>
  </button>
</div>
`;

exports[`Composant Card > Snapshots > devrait correspondre au snapshot avec des props visuelles 1`] = `
<div>
  <div
    class="flex flex-col relative overflow-hidden h-auto text-foreground box-border outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-large rounded-small w-full data-[hover=true]:bg-content2 dark:data-[hover=true]:bg-content2 bg-background/80 dark:bg-background/20 backdrop-blur-md backdrop-saturate-150 opacity-disabled cursor-not-allowed transition-transform-background motion-reduce:transition-none"
    data-disabled="true"
    tabindex="-1"
  >
    <div
      class="relative flex w-full p-3 flex-auto flex-col place-content-inherit align-items-inherit h-auto break-words text-left overflow-y-auto subpixel-antialiased"
    >
      Contenu de la Card
    </div>
  </div>
</div>
`;

exports[`Composant Card > Snapshots > devrait correspondre au snapshot avec les props par défaut 1`] = `
<div>
  <div
    class="flex flex-col relative overflow-hidden h-auto text-foreground box-border bg-content1 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium rounded-large transition-transform-background motion-reduce:transition-none"
    tabindex="-1"
  >
    <div
      class="flex p-3 z-10 w-full justify-start items-center shrink-0 overflow-inherit color-inherit subpixel-antialiased rounded-t-large"
    >
      En-tête de la Card
    </div>
    <div
      class="relative flex w-full p-3 flex-auto flex-col place-content-inherit align-items-inherit h-auto break-words text-left overflow-y-auto subpixel-antialiased"
    >
      Contenu de la Card
    </div>
    <div
      class="p-3 h-auto flex w-full items-center overflow-hidden color-inherit subpixel-antialiased rounded-b-large"
    >
      Pied de la Card
    </div>
  </div>
</div>
`;

exports[`Composant Card > Snapshots > devrait correspondre au snapshot sans en-tête ni pied 1`] = `
<div>
  <div
    class="flex flex-col relative overflow-hidden h-auto text-foreground box-border bg-content1 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium rounded-large transition-transform-background motion-reduce:transition-none"
    tabindex="-1"
  >
    <div
      class="relative flex w-full p-3 flex-auto flex-col place-content-inherit align-items-inherit h-auto break-words text-left overflow-y-auto subpixel-antialiased"
    >
      Contenu de la Card uniquement
    </div>
  </div>
</div>
`;
